### Introduction
A large number of microarray datasets in public repositories lack sample sex information. The absence of sample sex information restricts the reuse of 
gene expression data where the researcher intends to factor the effect of sex.

massiR is a package for predicting the sex of samples in microarray datasets.

### Mapping probes to genes


### Visual examination
```{r}
library(MetaIntegrator)
library(tidyverse)
geo21477 <- getGEOData("GSE21477")
keys <- geo21477$originalData$GSE21477$keys
x_probes <- keys[!is.na(keys) & keys == 'XIST']
# Expression data contains the expression levels of genes corresponding to certain probes
expData <- geo21477$originalData$GSE21477$expr
x_vals <- expData[names(x_probes),]
if (length(x_probes)> 1) { x_vals <- colMeans(x_vals, na.rm=TRUE)}

y_probes <- keys[!is.na(keys) & keys == 'RPS4Y1']
y_vals <- expData[names(y_probes),]
if (length(y_probes)> 1) { y_vals <- colMeans(y_vals, na.rm=TRUE)}
```

### Code for clustering
```{r}
ssProbes <- expData[c(names(x_probes), names(y_probes)), ] # extract the rows with the probes
cl <- kmeans(ssProbes, 2)

```

### Sex labeling with massiR
```{r}
biocLite("biomaRt")
require('biomaRt') # you might have to install this with Bioconductor (see Erika's instructions)
library(biomaRt)
ensembl.m <- useMart("ensembl") 
ensembl <- useDataset("hsapiens_gene_ensembl", mart=ensembl.m)
ychr.genes <- getBM(attributes= "hgnc_symbol",
                      filters=c("chromosome_name"),
                      values=list("Y"), mart=ensembl) 
```


### Taking a deeper dive into the sample metadata
```{r}
gse <- getGEOData("GSE50831")
pData <- gse$originalData$GSE50831$pheno
pData_factor <- apply(pData, 2, as.factor)
nlevelspercol <- apply(pData_factor, 2, nlevels)

```

### Sample sex-labeling script
Input: expression data frame
Output: data frame with sample IDs and sex labels


